onnxruntime/core/providers/rocm/reduction/reduction_ops.cc

```diff
@@ -0,0 +1,368 @@
+      kOnnxDomain,                                                              \
+      11, 11,                                                                   \
+      T,                                                                        \
+      kRocmExecutionProvider,                                                   \
+      KernelDefBuilder().TypeConstraint("T", DataTypeImpl::GetTensorType<T>()), \
+      name<T>);                                                                 \
+  ONNX_OPERATOR_TYPED_KERNEL_EX(                                                \
+      name,                                                                     \
+      kOnnxDomain,                                                              \
+      12,                                                                       \
+      T,                                                                        \
+      kRocmExecutionProvider,                                                   \
+      KernelDefBuilder().TypeConstraint("T", DataTypeImpl::GetTensorType<T>()), \
+      name<T>);
+
+static bool is_matrix_row_reduction(
+    const miopenReduceTensorOp_t miopen_reduce_op,
+    const int m,
+    const int n,
+    const size_t rank,
+    std::vector<int64_t> axes) {
+  if (m < 1)
+    return false;
+
+  if (n < 1)
+    return false;
+
+  if (rank < 2)
+    return false;
+
+  if (miopen_reduce_op != MIOPEN_REDUCE_TENSOR_ADD)
+    return false;
+
+  //empty axes, default reduction
+  if (axes.size() < 1)
+    return false;
+
+  return true;
+}
+
+// TODO ReduceKernel::ReduceKernelShared() is still used by some other training classes though it's not used here - this should be refactored.
+template <bool allow_multi_axes>
+template <typename T, typename OutT>
+Status ReduceKernel<allow_multi_axes>::ReduceKernelShared(
+    const T* X,
+    const TensorShape& input_shape,
+    OutT* Y,
+    const TensorShape& /*output_shape*/,
+    miopenReduceTensorOp_t miopen_reduce_op,
+    std::vector<int64_t> /*output_dims*/) const {
+  typedef typename ToHipType<T>::MappedType HipT;
+  const auto rank = input_shape.NumDimensions();
+
+// Block of fast matrix row reduction.
+// It relies on new atomicAdd for half type, so old hip can't use it.
+  const auto stride = input_shape[input_shape.NumDimensions() - 1];
+  const auto reduction_size = input_shape.Size() / stride;
+  if (fast_reduction_ && reduction_size <= std::numeric_limits<int>::max() && stride <= std::numeric_limits<int>::max() &&
+      is_matrix_row_reduction(miopen_reduce_op,
+        static_cast<int>(reduction_size),
+        static_cast<int>(stride), rank, axes_)) {
+
+    reduce_matrix_rows(
+      reinterpret_cast<const HipT*>(X),
+      reinterpret_cast<HipT*>(Y),
+      static_cast<int>(reduction_size),
+      static_cast<int>(stride));
+    return Status::OK();
+  }
+
+  return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "reduction1 is not supported");
```

Comment:
Can we make the error message more descriptive? What cases do reduction1/2/3 cover?

Reply:
There are some other code sections where miopen seems to replace cuda dnn. Is there a reason it couldn't be done here?

Reply:
OK, will add comment.